include/lin/substitutions.hpp

```diff
@@ -1,11 +1,273 @@
+namespace lin {
+namespace internal {
+
+/** @internal
+ *
+ *  @brief Tests whether or not the backward substitution algorithm can be
+ *         applied to a set of types.
+ *
+ *  @tparam C Upper triangular operator type.
+ *  @tparam D Unknown matrix or vector type.
+ *  @tparam E Know matrix or vector type.
+ *
+ *  @ingroup SUBSTITUTIONS
+ */
+template <class C, class D, class E>
+struct can_backward_sub : conjunction<
+    can_multiply<C, D>, have_same_elements<C, D, E>, have_same_dimensions<D, E>,
+    is_square<C>
+  > {};
+
+/** @internal
+ *
+ *  @brief Tests whether or not the forward substitution algorithm can be
+ *         applied to a set of types.
+ *
+ *  @tparam C Lower triangular operator type.
+ *  @tparam D Unknown matrix or vector type.
+ *  @tparam E Know matrix or vector type.
+ *
+ *  @ingroup SUBSTITUTIONS
+ */
+template <class C, class D, class E>
+struct can_forward_sub : can_backward_sub<C, D, E> {};
+
+} // namespace internal
+
+/** @brief Solves a linear system using backward substitution.
+ *
+ *  @tparam C
+ *  @tparam D
+ *  @tparam E
+ *
+ *  @param U Upper triangular operator.
+ *  @param X Unknown matrix or vector.
+ *  @param Y Know matrix or vector.
+ *
+ *  @return Zero on success and some other value otherwise.
+ *
+ *  Solves a linear system of the following form:
+ *
+ *  \f[
+ *    U X = Y
+ *  \f]
+ * 
+ *  where \f$U\f$ is an upper triangular operator, \f$X\f$ is an unknow matrix
+ *  or vector, and \f$Y\f$ is a known matrix or vector.
+ *
+ *  In order for the backward substitution to be valid, the matrix `U` must be
+ *  in upper triangular form. The function itself does not check for this and
+ *  simply assumes the elements below the main diagonal are all zero.
+ *
+ *  Lin assertion errors will be thrown if the dimensions of `X` are not the
+ *  same as the dimensions of `Y`.
+ *
+ *  @sa internal::can_backward_sub
+ *  @sa forward_sub
+ *
+ *  @ingroup SUBSTITUTIONS
+ */
+template <class C, class D, class E, typename =
+    std::enable_if_t<internal::can_backward_sub<C, D, E>::value>>
+constexpr size_t backward_sub(internal::Mapping<C> const &U, internal::Mapping<D> &X, internal::Mapping<E> const &Y) {
+  LIN_ASSERT(U.rows() == U.cols());
+  LIN_ASSERT(U.cols() == Y.rows());
+  LIN_ASSERT(Y.rows() == X.rows());
+  LIN_ASSERT(Y.cols() == X.cols());
+
+  // Solve for the last row
+  size_t m = U.rows() - 1;
+  row(X, m) = row(Y, m) / U(m, m);
+
+  // Solve for the other rows in descending order
+  size_t n = m - 1;
+  do {
+    typedef typename C::Traits::elem_t Elem;
+    typedef RowVector<Elem, 0, C::Traits::max_rows> T;
+    typedef std::conditional_t<internal::is_col_vector<E>::value, Vector<Elem, 0, E::Traits::max_rows>, Matrix<Elem, 0, E::Traits::cols, E::Traits::max_rows, E::Traits::max_cols>> V;
+
+    row(X, n) = (row(Y, n) - ref<T>(U, n, n + 1, m - n) * ref<V>(X, n + 1, 0, m - n, X.cols())) / U(n, n);
+  } while (n--);
+
+  // TODO : Change this to a meaningfull error code
+  return 0;
```

Comment:
noncritical?

Reply:
Yeah not really critical - at least not at the moment.